Add product develop document (#41)
voiio/voiio.github.io: product_development.md

# Product development

## Documents & Artifacts

### Kanban board

The Kanban board is the main tool of communication between stakeholders and the different product development units.

The Kanban board is owned by the product management team.

Engineers are responsible for maintaining cards on the Kanban board once they have pulled them into the backlog. Stories are always assigned when they are in progress, but never while they are in the backlog.

> All work that is done in either product, design or engineering should be reflected on the board. Stakeholders may have tickets on the board too.

Reviewer: ```suggestion
All work that is done in either product, design or engineering should be reflected on the board. Stakeholders may have tickets on the board too.

Priorities or urgencies are sorted top to bottom.
```

### OKR initiatives

> Strategic initiatives developed with a department and prioritized in close coordination with an objective owner. Those initiatives could be features and also research tasks, surveys, content work and more.

Reviewer 2: _**OKR initiatives are**_ strategic initiatives developed with ...

Reviewer: ```suggestion
Strategic initiatives are developed with a department and prioritized in close coordination with an objective owner. Those initiatives could be features and also research tasks, surveys, content work and more.
```

They are prioritized over anything else to ensure the success of our OKRs.

OKR initiatives should be in line with our long-term product vision.

> OKR initiatives are usually tracking in a separate lane on the Kanban board.

Reviewer 2: OKR initiatives are usually _**tracked**_ ...

Reviewer: ```suggestion
OKR initiatives are usually tracked in a separate lane on the Kanban board.
```

### New feature tickets

New features are the result of analysis or research based user demand.

> The product and design team owns the requirements engineering and conception of new features.

Reviewer 2: The product and design **_teams own_** the requirements ...

Reviewer: ```suggestion
The product and design teams own the requirements engineering and conception of new features.
```

> The product and design team owns the requirements engineering and conception of new features.

Reviewer 2: I actually don't know what "the requirements engineering" means.

Reviewer: Google it ;) https://en.wikipedia.org/wiki/Requirements_engineering

They are responsible for creating tickets that are ready to be pulled into the backlog.

### Growth opportunity

Anything that can help a department (not only product related) to improve their performance and the performance of our business.
They should be accompanied by metrics, their current state, and the projected state after the implementation of the growth opportunity.

E.g: Feature A will boost KPI B from 10% to 20% and will increase revenue by 10%.

Growth opportunities should align with our long term product vision.

Depending on the complexity those opportunities might be developed in small increments prioritizing value delivery.

> Growth opportunity are owned by a stakeholder and prioritized via the stakeholder meeting.

Reviewer 2: Growth **_opportunities_** are owned ...

Reviewer: ```suggestion
Growth opportunities are owned by a stakeholder and prioritized via the stakeholder meeting.
```

The product team is responsible to ensure that the ticket is ready to be pulled into the backlog.

### Operational hiccups

Operational hiccups are issues that disturb a stakeholders operation or hinder their ability to exercise their responsibilities.
They usually require some kind of technical support or facilitation and should have some level of urgency similar to bugs.

Operational hiccups are owned by a stakeholder and prioritized via the stakeholder meeting.
The product team is responsible to ensure that the ticket is ready to be pulled into the backlog.

### Bugs

Bugs are owned by the engineering team as they affect the platform's stability.

Bugs must have some kind of urgency. They are not prioritized by product, but triaged by engineering, based on what bug will kill our business first. Triage is handled by a designated senior engineer. Bugs are assigned to an engineer at all times but can be reassigned.

A bug ticket should have an average resolution rate of less than a week and no more than two weeks.

> Bugs may be closed or converted into a technical dept ticket, if there is no longer a high urgency and no impact on the stability of the platform.

Reviewer 2: Bugs may be closed or converted into a technical _**debt**_ ticket

Reviewer: ```suggestion
Bugs may be closed or converted into a technical debt ticket, if there is no longer a high urgency and no impact on the stability of the platform.
```

Updates on bug tickets, especially closure without resolution, need to be communicated with product proactively.

### Tech debt

Engineering leads are responsible for the stability of the platform and the development velocity, and consequently for when and what technical debt needs to be handled.

Work on technical debt must always be proactively communicated with the product team and documented in a ticket. Tech debt tickets must be pulled into the backlog in a refinement meeting, as they bind resources and take up significant space in the engineering backlog.

### Platform announcement Slack channel

Product, Design, Engineering, and Content team members post every change to go live on the platform into the Slack channel [#platform-announcements](https://voiio.slack.com/archives/C02BXK6R16V).

> This channel serves as a changelog for all employees and should include ever change the platform's experience, e.g.

Reviewer 2: ... and should include **_every_** change **_that affects_** the platform's experience ...

Reviewer: ```suggestion
This channel serves as a changelog for all employees and should include every change to the platform's experience, e.g.
```

* new features,
* bug fixes,
* content updates,
* etc.

## Meetings & Collaboration

### Agile principles

* Make work visible.
* Meaningful & actionable conversations.
* Don't ask for permission. Decisions are made, where the work is done.
* Regularly reflect on how we work together.

> ### Stakeholder Meetings

Reviewer 3: ```suggestion
### Stakeholder meetings
```

Stakeholder meetings are held weekly to discuss the product roadmap, upcoming features, and any blockers that need to be resolved or questions that need to be answered. It is the best path to discuss new ideas and to align on the product's direction. The product manager will relate work to the OKR progress and stakeholders will come prepared with issues/ideas and data, adding their topics to this document: [Stakeholder Meeting Platform](https://docs.google.com/document/d/1uhX68whx9T56unUlEUdyjUdq_dOJpHwO7Ycv892hqA8/edit)

The meeting is led by the Product Manager and attended by key stakeholders and an engineering lead.

> ### Refinement Meetings

Reviewer 3: ```suggestion
### Refinement meetings
```

# Product development

## Documents & Artifacts

### Kanban board

The Kanban board is the main tool of communication between stakeholders and the different product development units.

The Kanban board is owned by the product management team.

Engineers are responsible for maintaining cards on the Kanban board once they have pulled them into the backlog. Stories are always assigned when they are in progress, but never while they are in the backlog.

All work that is done in either product, design or engineering should be reflected on the board. Stakeholders may have tickets on the board too.

Priorities or urgencies are sorted top to bottom.

### OKR initiatives

Strategic initiatives are developed with a department and prioritized in close coordination with an objective owner. Those initiatives could be features and also research tasks, surveys, content work and more.

They are prioritized over anything else to ensure the success of our OKRs.

OKR initiatives should be in line with our long-term product vision.

OKR initiatives are usually tracked in a separate lane on the Kanban board.

### New feature tickets

> New features are the result of analysis or research based user demand.

Reviewer 4: ```suggestion
New features are the result of analysis or research based on the user demand.
```

Unless I got the sentence wrong - then it means, that it is a confusing sentence.

Reviewer 5: @codingjoe Grammarly schlägt mir Folgendes als richtig vor:
> New features result from analysis or research, based on user demand.

The product and design teams own the requirements engineering and conception of new features.

> They are responsible for creating tickets that are ready to be pulled into the backlog.

Reviewer: We were unsure. Isn't a new feature always a growth opportunity? I wonder if they need more distinction or should simply be merged.

Dailies are led by the respective team lead going through the Kanban board and attended by the entire team and, optionally, by people who have been closely working with the respective team.

### Retrospectives

Retrospectives are held every month with the engineering, product and design team. Stakeholders may join optionally or can be invited if they have been highly involved with a project.

### Review

We have a continuous development and deployment process and therefore no regular review meeting. However, reviews are an integral part of the development process. Engineers are responsible to include key stakeholders or product into the solution building and review process.

We strictly prioritize closing open work over everything else.
Therefore, reviewing and supporting ongoing projects is expected to be done within a business day.

Client-facing features may require a scheduled deployment coordinated with our clients by CSM.

### Development Kick-Off

> To prevent ticket ping pong, a kick-off is an optional short meeting, that can be scheduled by an engineer when they pick up new work. Its discussion should be reflected in the ticket. We aim for great tickets, but sometimes we miss.

Reviewer: This is new @karlawithakey, I'd love to get your opinion.

# Product development

## Documents & Artifacts

### Kanban board

The Kanban board is the main tool of communication between stakeholders and the different product development units.

The Kanban board is owned by the product management team.

Engineers are responsible for maintaining cards on the Kanban board once they have pulled them into the backlog. Stories are always assigned when they are in progress, but never while they are in the backlog.

All work that is done in either product, design or engineering should be reflected on the board. Stakeholders may have tickets on the board too.

### New feature tickets

The requirements engineering and conception of new features is owned by the product and design team.

### Growth opportunity

> Anything that can help a department (not only product related) to improve their performance and the performance of our business.

Author: I feel this should be better explained. E.g. with examples.

Growth opportunities should align with our long term product vision.

Depending on the complexity those opportunities might be developed in small increments prioritizing value delivery.

### OKR initiatives

> Strategic initiatives developed with a department and prioritized in close coordination with an objective owner. Those initiatives could be features and also research tasks, surveys, content work and more.

Author: It should be clear what OKR initiatives refer to in our day to day work. ...on the top lane of the roadmap board or something like that.

Reviewer: This might be subject to change. I believe we should mark all different ticket styles accordingly.

Reviewer: But I'll add a line for the lane.

They are prioritized over anything else to ensure the success of our OKRs.

OKR initiatives should be in line with our long-term product vision.

### Operational hiccups

> Operational hiccups are issues that disturb a stakeholders operation or hinder their ability to exercise their responsibilities. They usually require some kind of technical support or facilitation and should have some level of urgency similar to bugs but are outside of the engineering responsibilities.

Author: I'm not sure what this is and who is resolving it.

Reviewer 3: I am a bit lost in here...

1. How would the urgency be determined?  
2. Who is the responsible to resolve it? 

In many cases, each stakeholder have a different understanding and expectation of that is hindering their ability to exercise their responsibilities.

Reviewer: Good point, I will elaborate on that a bit more

### Bugs

Bugs are owned by the engineering team as they affect the platform's stability.

Bugs must have some kind of urgency. They are not prioritized by product, but triaged by engineering, based on what bug will kill our business first. Triage is handled by a designated senior engineer. Bugs are assigned to an engineer at all times but can be reassigned.

> A bug ticket should have an average resolution rate of less than a week.

Author: I'm not sure this is realistic. Especially with lower priority bugs. Except we consider lower priority bugs more important than OKR initiative work.

Reviewer 6: I think the idea is that if a bug is not important enough to be fixed in a week, either drop it completely or convert it to a story. That at least how I understood that section. Maybe it needs a slightly longer explanation here.

Reviewer: Yes, if it doesn't pose a short term risk, it may move to. But I don't want us to collect bugs, we are never going to fix.

Reviewer: Also, we are talking avg. here. Not max. I would set the max at 2 weeks before at bot comes in and sends a friendly reminder.

Bugs may be closed or converted into a story, if there is no longer a high urgency and no impact on the stability of the platform. Updates on bug tickets, especially closure without resolution, need to be communicated with product proactively.

### Tech debt

Engineering leads are responsible for the stability of the platform and the development velocity, and consequently for when and what technical debt needs to be handled.

Work on technical debt must always be proactively communicated with the product team and documented in a ticket. Tech debt tickets must be pulled into the backlog in a refinement meeting, as they bind resources and take up significant space in the engineering backlog.

> 

Author: Currently there is technical debt in our backlog but hardly anyone is touching it. It feels more like a "IF I have ever time to tackle this" instead of a "I could tackle this within the next week" thing.

Reviewer: Please bring that point to the engineering meeting. Maybe we can review tech dept ever meeting.

### Platform announcement Slack channel

> Product, Design, Engineering, and Content team members post every change to go live on the platform into the Slack channel [#platform-announcements](https://voiio.slack.com/archives/C02BXK6R16V). This channel serves as a changelog for all employees.

Author: I think this should be more precise. If there are changes to the platform (say we are improving the performance of a background sync job) we tend to not share this because there is no value in knowing this for anybody but the dev team and may make the channel too noisy.

Reviewer: This is working very well right now, I don't understand the confusion. Just keep doing, what you are doing and inform voiio of all changes to the platform experience.

## Meetings & Collaboration

### Agile principles

* Make work visible
* Meaningful & actionable conversations

> * Decisions are made, where the work is done

Author: I don't get this sentence

* Regularly reflect on how we work together

### Stakeholder Meetings

Stakeholder meetings are held weekly to discuss the product roadmap, upcoming features, and any blockers that need to be resolved or questions that need to be answered. It is the best path to discuss new ideas and to align on the product's direction. The product manager will relate work to the OKR progress and stakeholders will come prepared with issues/ideas and data, adding their topics to this document: [Stakeholder Meeting Platform](https://docs.google.com/document/d/1uhX68whx9T56unUlEUdyjUdq_dOJpHwO7Ycv892hqA8/edit)

The meeting is led by the Product Manager and attended by key stakeholders and an engineering lead.

### Refinement Meetings

Refinement meetings are held weekly and are the handover point from the Product Team to the Engineering Team. The product manager presents problems, opportunities, and priorities to the Engineering Team. Engineers ask questions and either pull the issue into the backlog or push it back to the product team for further refinement.

> The meeting is led by the product manager and attended by the product and engineering team.

Reviewer 4: Sometimes teams are capitalized (Product Team) and sometimes are not (product team). Let's align on the style. :)

Reviewer: AI ;) Should be all lowcaps

### Standups

Standups are held daily and are the heartbeat of the team (Engineering, Content, Product). They are used to update the team on the progress of ongoing tasks and blockers. Standups are not for problem-solving, but for identifying issues that need to be solved later during the day.

> Standups are led by the respective Team Lead going through the Kanban board and attended by the entire team and, optionally, by people who have been closely working with the respective team.

Author: A daily being led by someone is not a practice we follow. I'm not against it but we should reflect reality here.

Reviewer: That's OK as long as they work. However, if we loose discipline someone would need to facilitate them and be accountable. I wouldn't stress about it for now and this description isn't only for engineering dailies.

### Retrospectives

> Retrospectives are held every month with the Engineering, Product and Design Team. Stakeholders may join optionally or can be invited if they have been highly involved with a project.

Reviewer 4: We should mention that retroes are happening separately within the teams, not as a single big entity.

```suggestion
Retrospectives are held every month within the Engineering, Product, and Design teams. Stakeholders may join optionally or can be invited if they have been highly involved with a project.
```

### Review

We have a continuous development and deployment process and therefore no regular review meeting. However, reviews are an integral part of the development process. Engineers are responsible to include key stakeholders or product into the solution building and review process.

> We strictly prioritize closing open work over everything else. Therefore, any party that has been requested to review a feature is expected to make time within a business day.

Reviewer 4: ```suggestion
We place the highest priority on completing open work. As a result, any party asked to review a feature is expected to allocate time within one business day.
```

> Client-facing features may require a scheduled deployment coordinated with our clients by CSM.

Reviewer 4: May or may not - do we need to elaborate on this?

Reviewer: I honestly don't know yet. Let's see.
